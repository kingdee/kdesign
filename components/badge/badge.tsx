import React, { FunctionComponentElement, useContext, useRef, useMemo } from 'react'
import classNames from 'classnames'
import ConfigContext from '../config-provider/ConfigContext'
import { getCompProps } from '../_utils'
import { cloneElement } from '../_utils/reactNode'
import { tuple } from '../_utils/type'

export const PresetStatusColorTypes = tuple('success', 'processing', 'error', 'default', 'warning')
export type PresetStatusColorType = typeof PresetStatusColorTypes[number]

export const BadgeSizes = tuple('default', 'small')
export type BadgeSize = typeof BadgeSizes[number]
export interface RibbonProps {
  className?: string
  prefixCls?: string
  style?: React.CSSProperties
  text?: React.ReactNode
  // color?: LiteralUnion<PresetColorType, string>
  color?: string
  children?: React.ReactNode
}

export interface BadgeProps extends RibbonProps {
  count?: React.ReactNode
  showZero?: boolean
  overflowCount?: number
  dot?: boolean
  style?: React.CSSProperties
  prefixCls?: string
  scrollNumberPrefixCls?: string
  className?: string
  status?: PresetStatusColorType
  color?: string
  text?: React.ReactNode
  size?: BadgeSize
  offset?: [number | string, number | string]
  title?: string
}

const Badge = (props: BadgeProps): FunctionComponentElement<BadgeProps> => {
  const { getPrefixCls, prefixCls, compDefaultProps: userDefaultProps, direction } = useContext(ConfigContext)
  const badgeProps = getCompProps('Badge', userDefaultProps, props)
  const {
    prefixCls: customPrefixcls,
    children,
    status,
    text,
    color,
    count = null,
    overflowCount = 99,
    dot = false,
    size = 'default',
    title,
    offset,
    style,
    className,
    showZero = false,
    ...restProps
  } = badgeProps

  const badgePrefixCls = getPrefixCls!(prefixCls, 'badge', customPrefixcls) // 徽标数样式前缀
  const rtlCls = direction === 'rtl' ? `${badgePrefixCls}-rtl` : null
  // ================================ Misc ================================

  const numberedDisplayCount = count > overflowCount ? `${overflowCount}+` : count

  const hasStatus = (status !== null && status !== undefined) || (color !== null && color !== undefined)

  const isZero = numberedDisplayCount === '0' || numberedDisplayCount === 0

  const isSingleNumber = !count || typeof count !== 'object'

  const showAsDot = hasStatus || (!isZero && dot)

  const mergedCount = showAsDot ? '' : numberedDisplayCount

  const isHidden = useMemo(() => {
    const isEmpty = mergedCount === null || mergedCount === undefined || mergedCount === ''
    return (isEmpty || (isZero && !showZero)) && !showAsDot
  }, [mergedCount, isZero, showZero, showAsDot])

  const isDotRef = useRef(showAsDot)
  if (!isHidden) {
    isDotRef.current = showAsDot
  }

  // =============================== Styles ===============================
  const mergedStyle = useMemo<React.CSSProperties>(() => {
    if (!offset) {
      return { ...style }
    }
    const offsetStyle: React.CSSProperties = { marginTop: offset[1] }
    direction === 'rtl'
      ? (offsetStyle.left = -parseInt(offset[0] as string, 10))
      : (offsetStyle.right = -parseInt(offset[0] as string, 10))
    return {
      ...offsetStyle,
      ...style,
    }
  }, [offset, style])

  // =============================== Render ===============================

  const badgeClassName = classNames(
    badgePrefixCls,
    rtlCls,
    {
      [`${badgePrefixCls}-status`]: hasStatus,
      [`${badgePrefixCls}-not-a-wrapper`]: !children,
    },
    className,
  )

  const isDot = isDotRef.current

  const scrollNumberCls = classNames({
    [`${badgePrefixCls}-dot`]: isDot,
    [`${badgePrefixCls}-count`]: !isDot,
    [`${badgePrefixCls}-count-sm`]: size === 'small',
    [`${badgePrefixCls}-multiple-words`]: !isDot && mergedCount && mergedCount.toString().length > 1,
    [`${badgePrefixCls}-status-${status}`]: !!status,
    // [`${badgePrefixCls}-status-${color}`]: isPresetColor(color),
  })

  const statusCls = classNames({
    [`${badgePrefixCls}-status-dot`]: hasStatus,
    [`${badgePrefixCls}-status-${status}`]: !!status,
    // [`${prefixCls}-status-${color}`]: isPresetColor(color),
  })

  const statusStyle: React.CSSProperties = {}
  if (color) {
    statusStyle.background = color
  }

  // Title
  const titleNode = title ?? (typeof count === 'string' || typeof count === 'number' ? count : undefined)

  // Status Text
  const statusTextNode = isHidden || !text ? null : <span className={`${prefixCls}-status-text`}>{text}</span>

  // Display Component
  const displayNode = isSingleNumber ? (
    <sup className={scrollNumberCls} style={{ ...mergedStyle, backgroundColor: color }} title={titleNode}>
      {mergedCount}
    </sup>
  ) : (
    cloneElement(count, (oriProps: { style?: any; className?: any }) => ({
      style: {
        ...mergedStyle,
        ...oriProps.style,
      },
      className: classNames(`${badgePrefixCls}-custom-component`, oriProps?.className),
    }))
  )

  if (!children && hasStatus) {
    const statusTextColor = mergedStyle.color
    return (
      <span {...restProps} className={badgeClassName} style={mergedStyle}>
        {/* style={statusStyle} */}
        <span className={statusCls} style={statusStyle} />
        <span style={{ color: statusTextColor }} className={`${badgePrefixCls}-status-text`}>
          {text}
        </span>
      </span>
    )
  }

  return (
    <span {...restProps} className={badgeClassName}>
      {children}
      {!isHidden && displayNode}
      {statusTextNode}
    </span>
  )
}

Badge.displayName = 'Badge'
export default Badge
