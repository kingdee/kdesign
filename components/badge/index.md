---
category: Components
type: 展示
title: Badge
subtitle: 徽标数
order: 1
---
徽标是一种用于引导用户处理的组件，常见于图标或头像右上角。包含 2 种类型：
- 数字徽标
- 状态徽标

## 使用场景

需要引导用户关注内容时。
## API

| 属性 | 说明 | 类型 | 默认值 | 版本 |
| --- | --- | --- | --- | --- | --- |
| color | 自定义小圆点的颜色 | string | - | 1.0.0 |
| count | 展示的数字，大于 overflowCount 时显示为 `${overflowCount}+`，为 0 时隐藏 | ReactNode | - | 1.0.0 |
| dot | 不展示数字，只有一个小红点 | boolean | false | 1.0.0 |
| offset | 设置状态点的位置偏移 | \[number, number] | - | 1.0.0 |
| overflowCount | 展示封顶的数字值 | number | 99 | 1.0.0 |
| showZero | 当数值为 0 时，是否展示 Badge | boolean | false | 1.0.0 |
| size | 在设置了 `count` 的前提下有效，设置小圆点的大小 | `default` \| `small` | - | 1.0.0 |
| status | 设置 Badge 为状态点 | `success` \| `processing` \| `default` \| `error` \| `warning` | - | 1.0.0 |
| text | 在设置了 `status` 的前提下有效，设置状态点的文本 | ReactNode | - | 1.0.0 |
| title | 设置鼠标放在状态点上时显示的文字 | string | - | 1.0.0 |

## Design Token

| 分类 | 组件token | 全局token | 默认值 |
| --- | --- | --- | --- |
| color | --kd-c-badge-color-background | - | #fb2323 |
|  | --kd-c-badge-color | --kd-g-color-background | #fff |
|  | --kd-c-badge-color | --kd-g-color-text-primary | #212121 |
|  | --kd-c-badge-color-success | --kd-g-color-success | #1ba854 |
|  | --kd-c-badge-color-ongoing | --kd-g-color-ongoing | #276ff5 |
|  | --kd-c-badge-color-warning | --kd-g-color-warning | #ff991c |
|  | --kd-c-badge-color-error | --kd-g-color-error | #fb2323 |
|  | --kd-c-badge-color-default | - | #d9d9d9 |
| font | --kd-c-badge-font-weight | - | 400 |
|  | --kd-c-badge-font-size-base | --kd-g-font-size-middle | 14px |
|  | --kd-c-badge-font-size-small | --kd-g-font-size-small | 12px |
| sizing | --kd-c-badge-sizing-base | - | 16px |
|  | --kd-c-badge-sizing-small | - | 14px |
|  | --kd-c-badge-sizing-dot | - | 8px |
|  | --kd-c-badge-sizing-status | - | 8px |
