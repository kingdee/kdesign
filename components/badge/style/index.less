/*! rtl:begin:ignore */
@import '../../style/themes/index';
@import './mixin.less';
@badge-prefix-cls: ~'@{kd-prefix}-badge';
.@{badge-prefix-cls} {
  position: relative;
  display: inline-block;
  line-height: 1;
  font-size: @badge-font-size-base;

  &-count {
    z-index: auto;
    min-width: @badge-size-base;
    height: @badge-size-base;
    color: @badge-color-text;
    font-weight: @badge-font-weight;
    font-size: @font-size-small;
    line-height: @badge-size-base;
    white-space: nowrap;
    text-align: center;
    background: @badge-color-background;
    border-radius: calc(@badge-size-base / 2);
  }

  &-count-sm {
    min-width: @badge-size-small;
    height: @badge-size-small;
    padding: 0;
    font-size: @badge-font-size-small;
    line-height: @badge-size-small;
    border-radius: calc(@badge-size-small / 2);
  }

  &-dot {
    z-index: auto;
    width: @badge-size-dot;
    min-width: @badge-size-dot;
    height: @badge-size-dot;
    background: @badge-color-background;
    border-radius: 100%;
  }

  &-dot,
  &-custom-component,
  &-count {
    position: absolute;
    top: 0;
    right: 0;
    transform: translate(50%, -50%);
    transform-origin: 100% 0%;
  }

  &-multiple-words {
    padding: 0 4px;
  }

  &-status {
    line-height: inherit;
    vertical-align: baseline;
    &-text {
      margin-left: 8px;
      color: @badge-color-text-primary;
      font-size: @badge-font-size-base;
    }
    &-dot {
      position: relative;
      top: -1px;
      display: inline-block;
      width: @badge-size-status;
      height: @badge-size-status;
      vertical-align: middle;
      border-radius: 50%;
    }
    &-success {
      background-color: @badge-color-success;
    }
    &-processing {
      background-color: @badge-color-ongoing;
      &::after {
        position: absolute;
        top: -1px;
        left: -1px;
        width: 100%;
        height: 100%;
        border: 1px solid @badge-color-ongoing;
        border-radius: 50%;
        animation: antStatusProcessing 1.2s infinite ease-in-out;
        content: '';
      }
    }
    &-error {
      background-color: @badge-color-error;
    }
    &-default {
      background-color: @badge-color-default;
    }
    &-warning {
      background-color: @badge-color-warning;
    }
  }

  &-not-a-wrapper {
    &:not(.@{badge-prefix-cls}-status) {
      vertical-align: middle;
    }

    .@{badge-prefix-cls}-custom-component,
    .@{badge-prefix-cls}-count {
      position: relative;
      top: auto;
      display: block;
      transform-origin: 50% 50%;
      transform: none;
    }
  }
}

@keyframes antStatusProcessing {
  0% {
    transform: scale(0.8);
    opacity: 0.5;
  }
  100% {
    transform: scale(2.4);
    opacity: 0;
  }
}

.@{badge-prefix-cls}-rtl {
  direction: rtl;
  &.@{badge-prefix-cls} {
    .@{badge-prefix-cls}-dot,
    .@{badge-prefix-cls}-custom-component,
    .@{badge-prefix-cls}-count {
      position: absolute;
      left: 0;
      right: auto;
      transform: translate(-50%, -50%);
    }
  }
}
/*! rtl:end:ignore */
