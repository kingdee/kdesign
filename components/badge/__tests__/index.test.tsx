import React from 'react'
import { render, mount } from 'enzyme'
import Badge from '../index'
import ConfigProvider from '../../config-provider/index'
import { PresetStatusColorTypes, BadgeSizes } from '../badge'
import mountTest from '../../../tests/shared/mountTest'

describe('Badge', () => {
  // 1.mount test
  describe('1.mount test', () => {
    mountTest(Badge)
  })

  // 2.render test
  describe('2.render test', () => {
    it('renders correctly', () => {
      PresetStatusColorTypes.forEach((type) => {
        const wrapper = render(<Badge status={type} />)
        expect(wrapper).toMatchSnapshot()
      })
      BadgeSizes.forEach((size) => {
        const wrapper = render(<Badge count={10} size={size} />)
        expect(wrapper).toMatchSnapshot()
      })
    })
  })
  // 3.warns in component
  // 4. render null or undefined without errors
  describe('4.render null or undefined without errors', () => {
    it('render null or undefined Badge without errors', () => {
      expect(
        mount(
          <Badge>
            {null}
            {undefined}
          </Badge>,
        ),
      ).toBeEmptyRender()
    })
  })

  // 5. displayName
  describe('5.displayName', () => {
    it('should have displayName static property', () => {
      const wrapper = mount(<Badge>Text</Badge>)
      expect((wrapper.type() as any).displayName).toBe('Badge')
    })
  })

  // 6. class state
  describe('6.class state', () => {
    it('should class use right', () => {
      // default
      const DefaultBadge = mount(<Badge count={99} />)
      expect(DefaultBadge.find('.kd-badge-count').length).toBeGreaterThanOrEqual(1)

      // shhowZero
      const ShhowZeroBadge = mount(<Badge count={0} showZero />)
      expect(ShhowZeroBadge.find('.kd-badge-count').length).toBeGreaterThanOrEqual(1)

      // multiple count
      const MultipleBadge = mount(<Badge count={99} />)
      expect(MultipleBadge.find('.kd-badge-multiple-words').length).toBeGreaterThanOrEqual(1)

      // sm size
      const SizeSmBadge = mount(<Badge count={99} size="small" />)
      expect(SizeSmBadge.find('.kd-badge-count-sm').length).toBeGreaterThanOrEqual(1)

      // dot
      const DotBadge = mount(<Badge dot />)
      expect(DotBadge.find('.kd-badge-dot').length).toBeGreaterThanOrEqual(1)

      // status types
      PresetStatusColorTypes.forEach((type) => {
        const TestButton = mount(<Badge status={type} />)
        expect(TestButton.find(`.kd-badge-status-${type}`).length).toBeGreaterThanOrEqual(1)
      })
    })
  })

  // 7.component interaction(event)
  // 8.config provider
  describe('8.config provider', () => {
    it('should config use config provider', () => {
      const badgeConfig = {
        compDefaultProps: {
          Badge: {
            color: 'blue',
          },
        },
      }
      const wrapper = mount(
        <ConfigProvider value={badgeConfig}>
          <Badge>
            <span className="head-example" />
          </Badge>
        </ConfigProvider>,
      )
      expect(wrapper.find(`.kd-badge-dot`).prop('style')?.backgroundColor).toBe('blue')
    })
  })
  // 9. ref test
  // 10. api test
  describe('10.api test', () => {
    // color
    it('color', () => {
      const colorBadge = mount(
        <Badge color="blue">
          <span className="head-example" />
        </Badge>,
      )
      expect(colorBadge.find(`.kd-badge-dot`).prop('style')?.backgroundColor).toBe('blue')
    })
    // count
    it('count', () => {
      const countBadge = mount(
        <Badge count={5}>
          <span className="head-example" />
        </Badge>,
      )
      expect(countBadge.find(`.kd-badge-count`).prop('title')).toEqual(5)
      expect(countBadge.find(`.kd-badge-count`)).toHaveText('5')
    })
    // dot
    it('dot', () => {
      const dotBadge = mount(
        <Badge dot count={5}>
          <span className="head-example" />
        </Badge>,
      )
      expect(dotBadge.find(`.kd-badge-dot`).prop('title')).toEqual(5)
      expect(dotBadge.find(`.kd-badge-dot`)).toHaveText('')
    })
    // offset
    it('offset', () => {
      const offsetBadge = mount(
        <Badge count={5} offset={[10, 10]}>
          <span className="head-example" />
        </Badge>,
      )
      expect(offsetBadge.find(`.kd-badge-count`).prop('style')?.marginTop).toBe(10)
      expect(offsetBadge.find(`.kd-badge-count`).prop('style')?.right).toBe(-10)
    })
    // overflowCount
    it('overflowCount', () => {
      const overflowCountBadge = mount(
        <Badge count={100} overflowCount={10}>
          <span className="head-example" />
        </Badge>,
      )
      expect(overflowCountBadge.find(`.kd-badge-count`)).toHaveText('10+')
    })
    // showZero
    it('showZero', () => {
      const showZeroBadge = mount(
        <Badge showZero={false} count={0}>
          <span className="head-example" />
        </Badge>,
      )
      expect(showZeroBadge.exists('.kd-badge-count')).toBeFalsy()
    })
    // status
    it('status', () => {
      PresetStatusColorTypes.forEach((type) => {
        const statusBadge = mount(<Badge status={type} />)
        expect(statusBadge.exists('.kd-badge-count')).toBeFalsy()
      })
    })
    // text
    it('text', () => {
      PresetStatusColorTypes.forEach((type) => {
        const textBadge = mount(<Badge status={type} text={type} />)
        expect(textBadge.find(`.kd-badge-status-text`)).toHaveText(type)
      })
    })
    // title
    it('title', () => {
      const titleBadge = mount(
        <Badge count={100} overflowCount={10} title="300">
          <span className="head-example" />
        </Badge>,
      )
      expect(titleBadge.find(`.kd-badge-count`).prop('title')).not.toEqual(100)
      expect(titleBadge.find(`.kd-badge-count`).prop('title')).toEqual('300')
    })
  })
})
