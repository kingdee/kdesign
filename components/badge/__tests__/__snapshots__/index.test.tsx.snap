// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge 2.render test renders correctly 1`] = `
<span
  class="kd-badge kd-badge-status kd-badge-not-a-wrapper"
>
  <span
    class="kd-badge-status-dot kd-badge-status-success"
  />
  <span
    class="kd-badge-status-text"
  />
</span>
`;

exports[`Badge 2.render test renders correctly 2`] = `
<span
  class="kd-badge kd-badge-status kd-badge-not-a-wrapper"
>
  <span
    class="kd-badge-status-dot kd-badge-status-processing"
  />
  <span
    class="kd-badge-status-text"
  />
</span>
`;

exports[`Badge 2.render test renders correctly 3`] = `
<span
  class="kd-badge kd-badge-status kd-badge-not-a-wrapper"
>
  <span
    class="kd-badge-status-dot kd-badge-status-error"
  />
  <span
    class="kd-badge-status-text"
  />
</span>
`;

exports[`Badge 2.render test renders correctly 4`] = `
<span
  class="kd-badge kd-badge-status kd-badge-not-a-wrapper"
>
  <span
    class="kd-badge-status-dot kd-badge-status-default"
  />
  <span
    class="kd-badge-status-text"
  />
</span>
`;

exports[`Badge 2.render test renders correctly 5`] = `
<span
  class="kd-badge kd-badge-status kd-badge-not-a-wrapper"
>
  <span
    class="kd-badge-status-dot kd-badge-status-warning"
  />
  <span
    class="kd-badge-status-text"
  />
</span>
`;

exports[`Badge 2.render test renders correctly 6`] = `
<span
  class="kd-badge kd-badge-not-a-wrapper"
>
  <sup
    class="kd-badge-count kd-badge-multiple-words"
    title="10"
  >
    10
  </sup>
</span>
`;

exports[`Badge 2.render test renders correctly 7`] = `
<span
  class="kd-badge kd-badge-not-a-wrapper"
>
  <sup
    class="kd-badge-count kd-badge-count-sm kd-badge-multiple-words"
    title="10"
  >
    10
  </sup>
</span>
`;
